modules/tinymce/src/core/main/ts/fmt/FormatUtils.ts

```diff
@@ -18,39 +18,73 @@ const isNode = (node: any): node is Node => !!(node).nodeType;
 const isElementNode = (node: Node): node is Element =>
   NodeType.isElement(node) && !Bookmarks.isBookmarkNode(node) && !isCaretNode(node) && !NodeType.isBogus(node);
 
-const isInlineBlock = (node: Node): boolean => {
-  return node && /^(IMG)$/.test(node.nodeName);
+// Control blocks are any elements e.g. (img, iframe, noneditable element) that can be selected individually
+// In TinyMCE, selected blocks are indicated with the data-mce-selected attribute
+const isControlBlockSelected = (dom: DOMUtils, node: Node): boolean => {
+  // Table cells are a special case and are separately handled from native editor selection
+  if (isElementNode(node) && !/^(TD|TH)$/.test(node.nodeName)) {
+    const selectedAttr = dom.getAttrib(node, 'data-mce-selected');
+    const value = parseInt(selectedAttr, 10);
+    // Avoid cases where data-mce-selected is not a positive number e.g. inline-boundary
+    return !isNaN(value) && value > 0;
+  } else {
+    return false;
+  }
 };
 
 const isEditable = (elm: HTMLElement): boolean =>
   elm.isContentEditable === true;
 
-const moveStart = (dom: DOMUtils, selection: EditorSelection, rng: Range): void => {
-  const offset = rng.startOffset;
-  let container = rng.startContainer;
+const adjustSelectionAfter = <T>(editor: Editor, action: () => T, shouldMoveStart: (startNode: Node) => boolean): void => {
+  const { selection, dom } = editor;
+  const selectedNodeBeforeAction = selection.getNode();
+  const isSelectedBeforeNodeNoneditable = dom.getContentEditable(selectedNodeBeforeAction) === 'false';
 
-  if (container === rng.endContainer) {
-    if (isInlineBlock(container.childNodes[offset])) {
-      return;
-    }
+  action();
+
+  // Check previous selected node before the action still exists in the DOM
+  // and is still noneditable
+  const isBeforeNodeStillNoneditable = isSelectedBeforeNodeNoneditable && dom.getContentEditable(selectedNodeBeforeAction) === 'false';
+  if (isBeforeNodeStillNoneditable && dom.isChildOf(selectedNodeBeforeAction, editor.getBody())) {
+    editor.selection.select(selectedNodeBeforeAction);
+  } else if (shouldMoveStart(selection.getStart())) {
+    moveStartToNearestText(dom, selection);
   }
+};
+
+// Note: The reason why we only care about moving the start is because MatchFormat and its function use the start of the selection to determine if a selection has a given format or not
+// The format toolbar buttons use the MatchFormat functions to determine whether the button should be active or not
+// Therefore, if the start is not moved forward in certain cases, the match format detection and by extension,the toolbar button state will be incorrect
```

Comment:
I think the first note here is enough so I would remove this. It's how the formatter api works that the UI is using that api is besides the point.

Reply:
Yeah, you are probably right. Happy to remove those lines